pkg/sql/catalog/singleversion/svpreemptor/preemptor.go

```diff
@@ -0,0 +1,284 @@
+		g.GoCtx(func(ctx context.Context) (err error) {
+			outstandingLeases, err = p.checkForLeases(ctx, checkTS, descriptors)
+			return err
+		})
+		if err := g.Wait(); err != nil {
+			return err
+		}
+		log.Infof(ctx, "done with phases 1 and 2 %v: %d, %v", descriptors, len(outstandingLeases), outstandingLeases)
+
+		// Case 2b, there were no leases, and we got our locks.
+		if len(outstandingLeases) == 0 {
+			return nil
+		}
+
+		// Case 2a, our read of the leases was invalid because we got pushed writing
+		// our intents.
+		if txnTS := txn.ProvisionalCommitTimestamp(); !checkTS.Equal(txnTS) {
+			var err error
+			checkTS = txnTS
+			if outstandingLeases, err = p.checkForLeases(
+				ctx, checkTS, descriptors,
+			); err != nil {
+				return errors.Wrap(err, "refreshing the set of outstanding leases after push")
+			}
+		}
+	}
+
+	// Run phases 3 and 4 in parallel.
+	var timestampOfRemovals hlc.Timestamp
+	{
+		g := ctxgroup.WithContext(ctx)
+		toNotify := outstandingLeases.toIDSet()
+		g.GoCtx(func(ctx context.Context) error {
+			return p.notifyOutstandingLeases(ctx, toNotify)
+		})
+		g.GoCtx(func(ctx context.Context) (err error) {
+			err, timestampOfRemovals = p.waitForOutstandingLeases(
+				ctx, checkTS, outstandingLeases,
+			)
+			return err
+		})
+		// TODO(ajwerner): Detect error indicating that the transaction must be
+		// aborted and, well, abort the transaction or figure out if indeed it
+		// has been aborted or something.
+		if err := g.Wait(); err != nil {
+			return err
+		}
+	}
+
+	// Phase 5.
+	if log.ExpensiveLogEnabled(ctx, 1) {
+		log.Infof(ctx, "leases %d removed as of %v", descriptors, timestampOfRemovals)
+	}
+
+	return txn.ForwardWriteTimestamp(timestampOfRemovals)
+}
+
+func (p *Preemptor) writeLockingIntents(
+	ctx context.Context, txn *kv.Txn, descriptors catalog.DescriptorIDSet,
+) (err error) {
+	return p.s.Delete(ctx, txn, svstorage.ActionIDsToRows(svstorage.Lock, descriptors)...)
+}
+
+func (p *Preemptor) checkForLeases(
+	ctx context.Context, ts hlc.Timestamp, descriptors catalog.DescriptorIDSet,
+) (rowSet, error) {
+	ls := make(rowSet)
+	if err := p.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) (err error) {
+		ls.reset() // deal with restarts
+		if err := txn.SetFixedTimestamp(ctx, ts); err != nil {
+			return err
+		}
+		return p.s.Scan(ctx, txn, svstorage.ActionIDsToRows(svstorage.Lease, descriptors), ls.add)
+	}); err != nil {
+		return nil, err
+	}
+	return ls, nil
+}
+
+func (p *Preemptor) notifyOutstandingLeases(
+	ctx context.Context, toNotify catalog.DescriptorIDSet,
+) error {
+	return p.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		return p.s.Delete(ctx, txn, svstorage.ActionIDsToRows(svstorage.Notify, toNotify)...)
+	})
+}
+
+func (p *Preemptor) waitForOutstandingLeases(
+	ctx context.Context, startTS hlc.Timestamp, leases rowSet,
+) (_ error, leasesRemovedTimestamp hlc.Timestamp) {
+	idsOfInterest := leases.toIDSet()
+	removeExpired := func() error {
```

Comment:
not sure if this deserves a comment or a rename...Perhaps filterExpired might be better name?
Regardless, I want to make it clear that this does not remove rows from the leases table (not that I can see anyway); but simply ignores any  leases from sessions that are no longer valid, and thus we are guaranteed to converge to a state where there are no leases remaining.